### ISSUE TITLE
Error in dplyr::arrange, object 'qval' not found while comparing two seurat objects

### ISSUE THREAD

**Issue by Hainajin** [jackbibby1/SCPA#61]
Hi Jack,

I am trying to compare two seurat objects. I have previously manipulated the two objects from a larger dataset and the data has undergone normalization, pca, umap, and tsne.
I want to compare two subpopulations from the dataset with same cell type but different treatment group, so I extracted the data.

Here is my code for running SCPA:
pathways <- msigdbr("Mus musculus", "H") %>%
  format_pathways()
scpa_out <- compare_pathways(samples = list(b_cells_KA, b_cells_vnam), pathways = pathways, parallel = TRUE, cores = 4)

here is the error message:
Error in dplyr::arrange(., dplyr::desc(qval)) : 
Caused by error:
! object 'qval' not found

Thank you so much for your help!!
<img width="434" alt="1" src="https://github.com/jackbibby1/SCPA/assets/150500028/baada5f5-346e-4b70-ab79-6227f3994665">
<img width="532" alt="image" src="https://github.com/jackbibby1/SCPA/assets/150500028/720dd9ac-42a3-4973-b1b0-555cfed8dd8c">

**Comment by Hainajin** [comment]
I extracted their expression matrices and it worked!! Thank you so much Jack!

**Comment by st-tky** [comment]
Hi,
I have the same error message even after seurat_extract().
I use v5 Seurat and the latest version of SCPA (1.6.1).
Is there any reason for this error?
I would appreciate any suggestions.

**Comment by jackbibby1** [comment]
Hi. Can you send me the code you're running that gives you the error?

**Comment by st-tky** [comment]
Thank you for the quick response!
I am using my own seurat object.
Also, the code worked before but it's not now (I do not know the exact reasons for this).

The following is the code I am running:
resting <- seurat_extract(target_cells,
                          meta1 = "celltype", value_meta1 = "ABC",
                          meta2 = "state.ident", value_meta2 = "Steady",
                          assay = "SCT")

activated <- seurat_extract(target_cells,
                            meta1 = "celltype", value_meta1 = "ABC",
                            meta2 = "state.ident", value_meta2 = "Activated",
                            assay = "SCT" )

library(msigdbr)
pathways <- msigdbr("mouse", "H") %>% format_pathways()

rest_act <- compare_pathways(samples = list(resting, activated), 
                             pathways = pathways, parallel = TRUE, cores = 10)

## Error message ##
Processing in parallel using 10 cores
Cell numbers in population 1 = 13688
Cell numbers in population 2 = 250
- If greater than 500 cells, these populations will be downsampled
All 50 pathways passed the min/max genes threshold
Calculating pathway fold changes...
Performing a two-sample analysis with SCPA...
Error in `dplyr::arrange()`:
ℹ In argument: `..1 = qval`.
Caused by error:
! object 'qval' not found
Run `rlang::last_trace()` to see where the error occurred.
#################

**Comment by st-tky** [comment]
Thank you for the suggestion, Jack.
Unfortunately, I have the same error even with the code and dataset you suggested.
Could there be any version conflicts between packages?

#############
> library(SCPA); library(Seurat); library(dplyr); library(msigdbr)
> df <- readRDS("~/Downloads/naive_cd4.rds") # assuming this is where the downloaded dataset is
> df <- UpdateSeuratObject(df)
Validating object structure
Updating object slots
Ensuring keys are in the proper structure
Updating matrix keys for DimReduc ‘pca’
Updating matrix keys for DimReduc ‘umap’
Warning: Assay RNA changing from Assay to Assay
Warning: Assay integrated changing from Assay to Assay
Warning: Graph integrated_nn changing from Graph to Graph
Warning: Graph integrated_snn changing from Graph to Graph
Warning: DimReduc pca changing from DimReduc to DimReduc
Warning: DimReduc umap changing from DimReduc to DimReduc
Ensuring keys are in the proper structure
Ensuring feature names don't have underscores or pipes
Updating slots in RNA
Updating slots in integrated
Updating slots in integrated_nn
Setting default assay of integrated_nn to integrated
Updating slots in integrated_snn
Setting default assay of integrated_snn to integrated
Updating slots in pca
Updating slots in umap
Setting umap DimReduc to global
No assay information could be found for FindIntegrationAnchors
No assay information could be found for PairwiseIntegrateReference
Setting assay used for ScaleData.RNA to RNA
Setting assay used for ScaleData.integrated to integrated
Setting assay used for RunPCA.integrated to integrated
Setting assay used for RunUMAP.integrated.pca to integrated
Setting assay used for FindNeighbors.integrated.pca to integrated
No assay information could be found for FindClusters
Validating object structure for Assay ‘RNA’
Validating object structure for Assay ‘integrated’
Validating object structure for Graph ‘integrated_nn’
Validating object structure for Graph ‘integrated_snn’
Validating object structure for DimReduc ‘pca’
Validating object structure for DimReduc ‘umap’
Object representation is consistent with the most current Seurat version
Warning messages:
1: Adding a command log without an assay associated with it 
2: Adding a command log without an assay associated with it 
3: Adding a command log without an assay associated with it 
> resting <- seurat_extract(df,
+                           meta1 = "Cell_Type", value_meta1 = "Resting",
+                           meta2 = "Hour", value_meta2 = "0")
Extracting cells where Cell_Type == Resting AND Hour == 0
Extracting data from the RNA assay
> activated <- seurat_extract(df,
+                             meta1 = "Cell_Type", value_meta1 = "Activated",
+                             meta2 = "Hour", value_meta2 = "24")
Extracting cells where Cell_Type == Activated AND Hour == 24
Extracting data from the RNA assay
> pathways <- msigdbr("human", "H") %>% format_pathways()
> rest_act <- compare_pathways(samples = list(resting, activated), 
+                              pathways = pathways, 
+                              parallel = TRUE, 
+                              cores = 10)
Processing in parallel using 10 cores

Cell numbers in population 1 = 4341
Cell numbers in population 2 = 1697
- If greater than 500 cells, these populations will be downsampled

All 50 pathways passed the min/max genes threshold

Calculating pathway fold changes...

Performing a two-sample analysis with SCPA...
Error in `dplyr::arrange()`:
ℹ In argument: `..1 = qval`.
Caused by error:
! object 'qval' not found
Run `rlang::last_trace()` to see where the error occurred.

**Comment by jackbibby1** [comment]
Hmm, interesting. Yeah -- could you send me your `sessionInfo()` after running that script please?

**Comment by st-tky** [comment]
Yes, it is because the workflow worked before.
Here is the output of my sessionInfo():

R version 4.2.3 (2023-03-15)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS 14.3.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] msigdbr_7.5.1      dplyr_1.1.4        Seurat_5.0.1       SCPA_1.6.1         SeuratObject_5.0.1
[6] sp_1.6-0          

loaded via a namespace (and not attached):
  [1] Rtsne_0.16             colorspace_2.1-0       deldir_1.0-6           ellipsis_0.3.2        
  [5] ggridges_0.5.4         RcppHNSW_0.4.1         rstudioapi_0.14        spatstat.data_3.0-1   
  [9] leiden_0.4.3           listenv_0.9.0          ggrepel_0.9.3          RSpectra_0.16-1       
 [13] fansi_1.0.4            codetools_0.2-19       splines_4.2.3          doParallel_1.0.17     
 [17] polyclip_1.10-4        spam_2.10-0            jsonlite_1.8.4         ica_1.0-3             
 [21] cluster_2.1.4          png_0.1-8              uwot_0.1.14            shiny_1.7.5.1         
 [25] sctransform_0.4.1      spatstat.sparse_3.0-1  compiler_4.2.3         httr_1.4.5            
 [29] Matrix_1.6-5           fastmap_1.1.1          lazyeval_0.2.2         cli_3.6.1             
 [33] later_1.3.0            htmltools_0.5.5        tools_4.2.3            igraph_1.6.0          
 [37] dotCall64_1.1-1        gtable_0.3.3           glue_1.6.2             RANN_2.6.1            
 [41] reshape2_1.4.4         Rcpp_1.0.10            scattermore_1.2        vctrs_0.6.5           
 [45] babelgene_22.9         spatstat.explore_3.1-0 nlme_3.1-162           progressr_0.13.0      
 [49] iterators_1.0.14       lmtest_0.9-40          spatstat.random_3.1-4  stringr_1.5.0         
 [53] globals_0.16.2         mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.3       
 [57] irlba_2.3.5.1          goftest_1.2-3          future_1.33.0          MASS_7.3-58.3         
 [61] zoo_1.8-12             scales_1.2.1           promises_1.2.0.1       spatstat.utils_3.0-2  
 [65] parallel_4.2.3         RColorBrewer_1.1-3     reticulate_1.28        pbapply_1.7-0         
 [69] gridExtra_2.3          ggplot2_3.4.2          stringi_1.7.12         fastDummies_1.7.3     
 [73] foreach_1.5.2          rlang_1.1.0            pkgconfig_2.0.3        matrixStats_0.63.0    
 [77] lattice_0.21-8         ROCR_1.0-11            purrr_1.0.1            tensor_1.5            
 [81] patchwork_1.2.0        htmlwidgets_1.6.2      cowplot_1.1.1          tidyselect_1.2.0      
 [85] parallelly_1.35.0      RcppAnnoy_0.0.20       plyr_1.8.9             magrittr_2.0.3        
 [89] R6_2.5.1               generics_0.1.3         pillar_1.9.0           withr_2.5.0           
 [93] fitdistrplus_1.1-8     survival_3.5-5         abind_1.4-5            tibble_3.2.1          
 [97] future.apply_1.10.0    KernSmooth_2.23-20     utf8_1.2.3             spatstat.geom_3.1-0   
[101] plotly_4.10.1          grid_4.2.3             data.table_1.14.9      digest_0.6.31         
[105] xtable_1.8-4           tidyr_1.3.0            httpuv_1.6.9           munsell_0.5.0         
[109] viridisLite_0.4.1

**Comment by jackbibby1** [comment]
Yeah, I'm struggling to replicate this from my side. I've tried with a couple of R versions and a fresh installation, and things run smoothly on my end. You're running an older version of R (v4.2.3) and sp (v1.6.0), but I'm not sure this should be an issue. The only other thing I can think of is uninstalling SCPA, Seurat and tidyverse, and reinstalling them, and making sure all your packages are up date after that with `update.packages()`

**Comment by st-tky** [comment]
In accordance with your suggestion, I have installed the latest version of R (v4.3.3) and packages, which has solved my problem!
Thank you so much for your kind help!
Also, I should have tried this before asking, sorry to bother you.
I hope these working around will be some help for other users loving this nice package.
I appreciate your kind assistance and this informative package!

Just for your information, the following is the updated sessionInfo().
> sessionInfo()
R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.3.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] msigdbr_7.5.1      dplyr_1.1.4        Seurat_5.0.3       SCPA_1.6.1         SeuratObject_5.0.1 sp_2.1-3          

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     shape_1.4.6.1          rstudioapi_0.16.0      jsonlite_1.8.8         magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1           GlobalOptions_0.1.2   
  [9] fs_1.6.3               vctrs_0.6.5            ROCR_1.0-11            memoise_2.0.1          spatstat.explore_3.2-7 htmltools_0.5.7        usethis_2.2.3          curl_5.2.1            
 [17] sctransform_0.4.1      parallelly_1.37.1      KernSmooth_2.23-22     htmlwidgets_1.6.4      desc_1.4.3             ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [25] zoo_1.8-12             cachem_1.0.8           igraph_2.0.3           mime_0.12              lifecycle_1.0.4        iterators_1.0.14       pkgconfig_2.0.3        Matrix_1.6-5          
 [33] R6_2.5.1               fastmap_1.1.1          clue_0.3-65            fitdistrplus_1.1-11    future_1.33.1          shiny_1.8.0            digest_0.6.35          colorspace_2.1-0      
 [41] S4Vectors_0.40.2       patchwork_1.2.0        ps_1.7.6               tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1          pkgload_1.3.4          labeling_0.4.3        
 [49] progressr_0.14.0       fansi_1.0.6            spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6        abind_1.4-5            compiler_4.3.3         remotes_2.5.0         
 [57] doParallel_1.0.17      withr_3.0.0            fastDummies_1.7.3      pkgbuild_1.4.4         MASS_7.3-60.0.1        sessioninfo_1.2.2      rjson_0.2.21           tools_4.3.3           
 [65] lmtest_0.9-40          httpuv_1.6.14          future.apply_1.11.1    goftest_1.2-3          glue_1.7.0             callr_3.7.5            nlme_3.1-164           promises_1.2.1        
 [73] grid_4.3.3             Rtsne_0.17             cluster_2.1.6          reshape2_1.4.4         generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-4    tidyr_1.3.1           
 [81] data.table_1.15.2      utf8_1.2.4             BiocGenerics_0.48.1    spatstat.geom_3.2-9    RcppAnnoy_0.0.22       ggrepel_0.9.5          RANN_2.6.1             foreach_1.5.2         
 [89] pillar_1.9.0           stringr_1.5.1          babelgene_22.9         spam_2.10-0            RcppHNSW_0.6.0         later_1.3.2            circlize_0.4.16        splines_4.3.3         
 [97] lattice_0.22-6         survival_3.5-8         deldir_2.0-4           tidyselect_1.2.1       ComplexHeatmap_2.18.0  miniUI_0.1.1.1         pbapply_1.7-2          gridExtra_2.3         
[105] IRanges_2.36.0         scattermore_1.2        stats4_4.3.3           devtools_2.4.5         matrixStats_1.2.0      stringi_1.8.3          lazyeval_0.2.2         codetools_0.2-19      
[113] tibble_3.2.1           BiocManager_1.30.22    cli_3.6.2              uwot_0.1.16            xtable_1.8-4           reticulate_1.35.0      munsell_0.5.0          processx_3.8.4        
[121] Rcpp_1.0.12            globals_0.16.3         spatstat.random_3.2-3  png_0.1-8              parallel_4.3.3         ellipsis_0.3.2         ggplot2_3.5.0          dotCall64_1.1-1       
[129] profvis_0.3.8          urlchecker_1.0.1       listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0           ggridges_0.5.6         crayon_1.5.2           leiden_0.4.3.1

**Comment by jackbibby1** [comment]
Great, glad it's sorted then. Thanks for the update

Jack

**Comment by jackbibby1** [comment]
I'm on vacation at the minute but I'll change this in a couple of weeks. Specifying cores should be good for now

**Comment by emmanuel-contreras** [comment]
I am running into this error ```! object 'qval' not found``` on a cloud environment even after specifying `parallel=T` and specifying the number of cores. Setting parallel to False works but processing is slower
